Utilizando esse conjunto de dados:
https://www.kaggle.com/datasets/disham993/9000-movies-dataset

Responda as seguintes perguntas:

1.   Qual tamanho do DataSet?
2.   Quantas linhas?
3.   Quantas colunas?
4.   Qual o tipo de variável de cada coluna?
5.   Qual o filme com maior número de votaçôes?
6.   Qual filme teve a maior nota (critério de desempate é o filme com mais votos)
7.   Existem valores nulos? Se sim, qual tratamento irá realizar? (Se não temos nome de algum filme, melhor nem considerar)
8.   Transforme as variaveis categóricas de linguagem e genero em númericas (utilize dummy)
9.   Normalize as variaveis numéricas
10.  Armazene esses valores como um artefato dentro do MLFlow
11. QUais insights é possivel obter desses dados?



In [ ]:
#Instalação do MLFlow
!pip install mlflow

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.8/8.8 MB 76.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.3/2.3 MB 109.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 86.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 147.8/147.8 kB 15.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.9/114.9 kB 14.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 85.0/85.0 kB 10.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 752.6/752.6 kB 60.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 203.4/203.4 kB 21.6 MB/s eta 0:00:00


In [ ]:
#Manipulação de Dados
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import mlflow

In [ ]:
#Carregar Dataset
dados = pd.read_csv("/content/mymoviedb.csv", engine='python')
dados.head()

Release_Date                    Title  \
0   2021-12-15  Spider-Man: No Way Home   
1   2022-03-01               The Batman   
2   2022-02-25                  No Exit   
3   2021-11-24                  Encanto   
4   2021-12-22           The King's Man   

                                            Overview  Popularity Vote_Count  \
0  Peter Parker is unmasked and no longer able to...    5083.954       8940   
1  In his second year of fighting crime, Batman u...    3827.658       1151   
2  Stranded at a rest stop in the mountains durin...    2618.087        122   
3  The tale of an extraordinary family, the Madri...    2402.201       5076   
4  As a collection of history's worst tyrants and...    1895.511       1793   

  Vote_Average Original_Language                               Genre  \
0          8.3                en  Action, Adventure, Science Fiction   
1          8.1                en            Crime, Mystery, Thriller   
2          6.3                en                            Thriller   
3          7.7                en  Animation, Comedy, Family, Fantasy   
4          7.0                en    Action, Adventure, Thriller, War   

                                          Poster_Url  
0  https://image.tmdb.org/t/p/original/1g0dhYtq4i...  
1  https://image.tmdb.org/t/p/original/74xTEgt7R3...  
2  https://image.tmdb.org/t/p/original/vDHsLnOWKl...  
3  https://image.tmdb.org/t/p/original/4j0PNHkMr5...  
4  https://image.tmdb.org/t/p/original/aq4Pwv5Xeu...

In [ ]:
import os

In [ ]:
print(dados.info())
size = os.path.getsize("/content/mymoviedb.csv")
print(size/(1024*1024))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9837 entries, 0 to 9836
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Release_Date       9837 non-null   object 
 1   Title              9828 non-null   object 
 2   Overview           9828 non-null   object 
 3   Popularity         9827 non-null   float64
 4   Vote_Count         9827 non-null   object 
 5   Vote_Average       9827 non-null   object 
 6   Original_Language  9827 non-null   object 
 7   Genre              9826 non-null   object 
 8   Poster_Url         9826 non-null   object 
dtypes: float64(1), object(8)
memory usage: 691.8+ KB
None
4.013148307800293


## 1. Qual tamanho do DataSet?
4Mb

## 2. Quantas linhas?
9837 Registros

##3. Quantas colunas?
9 Colunas

##4. Qual o tipo de variável de cada coluna?
Coluna    -   tipo <br>
  Release_Date - object <br>
  Title - object <br>
  Overview - object <br>
  Popularity - float64 <br>
  Vote_Count - object <br>
  Vote_Average - object <br>
  Original_Language - object <br>
  Genre - object <br>
  Poster_Url - object                                                 

In [ ]:
display(dados.head())

Release_Date                    Title  \
0   2021-12-15  Spider-Man: No Way Home   
1   2022-03-01               The Batman   
2   2022-02-25                  No Exit   
3   2021-11-24                  Encanto   
4   2021-12-22           The King's Man   

                                            Overview  Popularity Vote_Count  \
0  Peter Parker is unmasked and no longer able to...    5083.954          u   
1  In his second year of fighting crime, Batman u...    3827.658          u   
2  Stranded at a rest stop in the mountains durin...    2618.087          u   
3  The tale of an extraordinary family, the Madri...    2402.201          u   
4  As a collection of history's worst tyrants and...    1895.511          u   

  Vote_Average Original_Language                               Genre  \
0          8.3                en  Action, Adventure, Science Fiction   
1          8.1                en            Crime, Mystery, Thriller   
2          6.3                en                            Thriller   
3          7.7                en  Animation, Comedy, Family, Fantasy   
4          7.0                en    Action, Adventure, Thriller, War   

                                          Poster_Url  
0  https://image.tmdb.org/t/p/original/1g0dhYtq4i...  
1  https://image.tmdb.org/t/p/original/74xTEgt7R3...  
2  https://image.tmdb.org/t/p/original/vDHsLnOWKl...  
3  https://image.tmdb.org/t/p/original/4j0PNHkMr5...  
4  https://image.tmdb.org/t/p/original/aq4Pwv5Xeu...

In [ ]:
# Converte coluna para numerico
dados['Vote_Count'] = pd.to_numeric(dados['Vote_Count'], errors='coerce')
dados['Vote_Average'] = pd.to_numeric(dados['Vote_Average'], errors='coerce')

lista_nota = []
maior_nota = dados['Vote_Average'].max()
print(maior_nota)

filmes_com_maior_nota = dados[dados['Vote_Average'] == maior_nota]
print(filmes_com_maior_nota)
filme_final = dados.loc[filmes_com_maior_nota['Vote_Average'].idxmax()]

if len(filmes_com_maior_nota) > 1:

    filme_final = filmes_com_maior_nota.loc[filmes_com_maior_nota['Vote_Count'].idxmax()]

display(filme_final)

10.0
     Release_Date                       Title  \
9401   2020-04-09  Kung Fu Master Huo Yuanjia   

                                               Overview  Popularity  \
9401  The young and vigorous Huo Yuanjia was only fo...      13.745   

      Vote_Count  Vote_Average Original_Language          Genre  \
9401         NaN          10.0                zh  Action, Drama   

                                             Poster_Url  
9401  https://image.tmdb.org/t/p/original/boXAHksMko...  


Release_Date                                                2020-04-09
Title                                       Kung Fu Master Huo Yuanjia
Overview             The young and vigorous Huo Yuanjia was only fo...
Popularity                                                      13.745
Vote_Count                                                         NaN
Vote_Average                                                      10.0
Original_Language                                                   zh
Genre                                                    Action, Drama
Poster_Url           https://image.tmdb.org/t/p/original/boXAHksMko...
Name: 9401, dtype: object

##5. Qual o filme com maior número de votações?

A coluna 'Vote_Count' vem com valores incorretos para análise

##6. Qual filme teve a maior nota (critério de desempate é o filme com mais votos)

##7. Existem valores nulos? Se sim, qual tratamento irá realizar? (Se não temos nome de algum filme, melhor nem considerar)

In [ ]:
display(dados.isnull().sum().sort_values(ascending=False))

Vote_Count           9826
Release_Date            0
Title                   0
Overview                0
Popularity              0
Vote_Average            0
Original_Language       0
Genre                   0
Poster_Url              0
dtype: int64

##8. Transforme as variaveis categóricas de linguagem e genero em númericas (utilize dummy)

In [ ]:
dummies_linguagem = pd.get_dummies(dados['Original_Language'], prefix='Lang', drop_first=True)

dummies_genero = dados['Genre'].str.get_dummies(sep=',')

# 'axis=1' significa que estamos concatenando colunas
dados_numericos = pd.concat([dados, dummies_linguagem, dummies_genero], axis=1)

# Remove as colunas de texto originais que não são mais necessárias
dados_numericos = dados_numericos.drop(['Original_Language', 'Genre'], axis=1)

# Mostra as primeiras linhas do resultado
print("DataFrame com variáveis Dummies:")
display(dados_numericos.head())

DataFrame com variáveis Dummies:


Release_Date                    Title  \
0   2021-12-15  Spider-Man: No Way Home   
1   2022-03-01               The Batman   
2   2022-02-25                  No Exit   
3   2021-11-24                  Encanto   
4   2021-12-22           The King's Man   

                                            Overview  Popularity  Vote_Count  \
0  Peter Parker is unmasked and no longer able to...    5083.954         NaN   
1  In his second year of fighting crime, Batman u...    3827.658         NaN   
2  Stranded at a rest stop in the mountains durin...    2618.087         NaN   
3  The tale of an extraordinary family, the Madri...    2402.201         NaN   
4  As a collection of history's worst tyrants and...    1895.511         NaN   

   Vote_Average                                         Poster_Url  Lang_bn  \
0           8.3  https://image.tmdb.org/t/p/original/1g0dhYtq4i...    False   
1           8.1  https://image.tmdb.org/t/p/original/74xTEgt7R3...    False   
2           6.3  https://image.tmdb.org/t/p/original/vDHsLnOWKl...    False   
3           7.7  https://image.tmdb.org/t/p/original/4j0PNHkMr5...    False   
4           7.0  https://image.tmdb.org/t/p/original/aq4Pwv5Xeu...    False   

   Lang_ca  Lang_cn  ...  History  Horror  Music  Mystery  Romance  \
0    False    False  ...        0       0      0        0        0   
1    False    False  ...        0       0      0        0        0   
2    False    False  ...        0       0      0        0        0   
3    False    False  ...        0       0      0        0        0   
4    False    False  ...        0       0      0        0        0   

   Science Fiction  TV Movie  Thriller  War  Western  
0                0         0         0    0        0  
1                0         0         0    0        0  
2                0         0         1    0        0  
3                0         0         0    0        0  
4                0         0         0    0        0  

[5 rows x 87 columns]

##9. Normalize as variaveis numéricas

In [ ]:
from sklearn.preprocessing import StandardScaler
colunas_para_escalar = ['Vote_Count', 'Vote_Average']

# Copia
df_padronizado = dados_numericos.copy()

# 2. Inicializa o scaler
scaler_std = StandardScaler()

# 3. Padroniza numéricas
df_padronizado[colunas_para_escalar] = scaler_std.fit_transform(df_padronizado[colunas_para_escalar])

# 4. Verifique o resultado
print("\n--- DataFrame Padronizado (Standard) ---")
display(df_padronizado[colunas_para_escalar].describe())
# display(df_padronizado.head())

#Salvar
dados_processados = df_padronizado.copy()
dados_processados.to_csv('dados_processados.csv', index=False)
print("Dataset processado salvo localmente.")


--- DataFrame Padronizado (Standard) ---


/usr/local/lib/python3.12/dist-packages/sklearn/utils/extmath.py:1101: RuntimeWarning: invalid value encountered in divide
  updated_mean = (last_sum + new_sum) / updated_sample_count
/usr/local/lib/python3.12/dist-packages/sklearn/utils/extmath.py:1106: RuntimeWarning: invalid value encountered in divide
  T = new_sum / new_sample_count
/usr/local/lib/python3.12/dist-packages/sklearn/utils/extmath.py:1126: RuntimeWarning: invalid value encountered in divide
  new_unnormalized_variance -= correction**2 / new_sample_count


Vote_Count  Vote_Average
count         0.0  9.826000e+03
mean          NaN  5.524676e-16
std           NaN  1.000051e+00
min           NaN -5.699957e+00
25%           NaN -4.775143e-01
50%           NaN  5.358163e-02
75%           NaN  5.846775e-01
max           NaN  3.151641e+00

Dataset processado salvo localmente.


##10. Armazene esses valores como um artefato dentro do **MLFlow**

In [ ]:
# Salvar o dataset processado como um arquivo CSV
dados_processados = "dados_processados.csv"
dados.to_csv(dados_processados, index=False)
print("Dataset processado salvo localmente.")

# Registrar o dataset processado como artefato no MLflow
mlflow.start_run()  # Iniciar um novo experimento
mlflow.log_artifact(dados_processados)  # Registrar o arquivo como artefato
mlflow.end_run()  # Encerrar o experimento

print("Features armazenadas e versionadas com sucesso no MLflow!")

Dataset processado salvo localmente.
Features armazenadas e versionadas com sucesso no MLflow!


##11. QUais insights é possivel obter desses dados?

In [ ]:
#Quais Generos são mais populares?
media_popularidade = dados.groupby('Genre')['Popularity'].mean()
print("Média de Popularidade por Gênero:")
print(media_popularidade.sort_values(ascending=False).head(10))

#Qual o filme mais popular?
filme_mais_popular = dados.loc[dados['Popularity'].idxmax()]
print("\nFilme Mais Popular:")
print(filme_mais_popular)

Média de Popularidade por Gênero:
Genre
Animation, Family, Fantasy, Comedy, Adventure    1373.778000
Action, Crime, Fantasy                            841.059500
Animation, Comedy, Family, Music                  626.981500
Animation, Comedy, Family, Western                614.110000
Animation, Drama, Comedy, Horror                  551.650000
Animation, Science Fiction, Family, Comedy        509.277000
Animation, Action, Fantasy, Adventure             453.796500
Comedy, Action, Adventure, Science Fiction        447.872000
Animation, Comedy, Family, Fantasy                430.496375
Fantasy, Comedy, Adventure                        428.540000
Name: Popularity, dtype: float64

Filme Mais Popular:
Release_Date                                                2021-12-15
Title                                          Spider-Man: No Way Home
Overview             Peter Parker is unmasked and no longer able to...
Popularity                                                    5083.954
Vote_Count  